src/mbgl/util/bounding_volumes.cpp

```diff
@@ -0,0 +1,283 @@
+        vec3i{far_br, far_tr, far_tl}     // far
+    };
+
+    if (flippedY) {
+        std::for_each(frustumPlanePointIndices.begin(), frustumPlanePointIndices.end(), [](vec3i& tri) {
+            std::swap(tri[1], tri[2]);
+        });
+    }
+
+    std::array<vec4, 6> frustumPlanes;
+
+    for (std::size_t i = 0; i < frustumPlanePointIndices.size(); i++) {
+        const vec3i indices = frustumPlanePointIndices[i];
+
+        // Compute plane equation using 3 points on the plane
+        const vec3 p0 = toVec3(cornerCoords[indices[0]]);
+        const vec3 p1 = toVec3(cornerCoords[indices[1]]);
+        const vec3 p2 = toVec3(cornerCoords[indices[2]]);
+
+        const vec3 a = vec3Sub(p0, p1);
+        const vec3 b = vec3Sub(p2, p1);
+        const vec3 n = vec3Normalize(vec3Cross(a, b));
+
+        frustumPlanes[i] = {n[0], n[1], n[2], -vec3Dot(n, p1)};
+    }
+
+    std::array<vec3, 8> frustumPoints;
+
+    for (size_t i = 0; i < cornerCoords.size(); i++) frustumPoints[i] = toVec3(cornerCoords[i]);
+
+    return {frustumPoints, frustumPlanes};
+}
+
+IntersectionResult Frustum::intersects(const AABB& aabb) const {
+    // Execute separating axis test between two convex objects to find intersections
+    // Each frustum plane together with 3 major axes define the separating axes
+    // This implementation is conservative as it's not checking all possible axes.
+    // False positive rate is ~0.5% of all cases (see intersectsPrecise).
+    // Note: test only 4 points as both min and max points have zero elevation
+    assert(aabb.min[2] == 0.0 && aabb.max[2] == 0.0);
+
+    if (!bounds.intersects(aabb)) return IntersectionResult::Separate;
+
+    const std::array<vec4, 4> aabbPoints = {
+        vec4{aabb.min[0], aabb.min[1], 0.0, 1.0},
+        vec4{aabb.max[0], aabb.min[1], 0.0, 1.0},
+        vec4{aabb.max[0], aabb.max[1], 0.0, 1.0},
+        vec4{aabb.min[0], aabb.max[1], 0.0, 1.0},
+    };
+
+    bool fullyInside = true;
+
+    for (const vec4& plane : planes) {
+        size_t pointsInside = 0;
+
+        pointsInside += vec4Dot(plane, aabbPoints[0]) >= 0.0;
+        pointsInside += vec4Dot(plane, aabbPoints[1]) >= 0.0;
+        pointsInside += vec4Dot(plane, aabbPoints[2]) >= 0.0;
+        pointsInside += vec4Dot(plane, aabbPoints[3]) >= 0.0;
+
+        if (!pointsInside) {
+            // Separating axis found, no intersection
+            return IntersectionResult::Separate;
+        }
+
+        if (pointsInside != aabbPoints.size()) fullyInside = false;
+    }
+
+    return fullyInside ? IntersectionResult::Contains : IntersectionResult::Intersects;
+}
+
+IntersectionResult Frustum::intersectsPrecise(const AABB& aabb, bool edgeCasesOnly) const {
+    if (!edgeCasesOnly) {
+        IntersectionResult result = intersects(aabb);
+
+        if (result == IntersectionResult::Separate) return result;
+    }
+
+    const std::array<vec3, 4> aabbPoints = {vec3{aabb.min[0], aabb.min[1], 0.0},
```

Comment:
Could alloc and copy here get removed, and add methods to access min and max also for dim 2?

Reply:
All these arrays are stack allocated so the only cost should be variable initialization.

Z-dimension is redundant as all tiles are on the xy-plane and leaving it out reduces the number of required intersection tests. This could be updated if we ever have a need for elevation.

Corner points of the aabb could be pre-generated but it would be premature optimization in this case. Intersection tests are performed once per aabb and moving the generation of points to constructor wouldn't remove any logic. `Frustum::intersects` also requires a `w` component whereas `Frustum::intersectsPrecise` does not.